packages/mds-trip-processor/src/proc-trip.ts

```diff
@@ -0,0 +1,156 @@
+    return null
+  }
+
+  // Calculate event binned trip telemetry data
+  const tripMap = await cache.readTripsTelemetry(`${provider_id}:${device_id}`)
+  if (tripMap) {
+    // Get trip metadata
+    const tripStartEvent = events[0]
+    const tripEndEvent = events[events.length - 1]
+    const baseTripData = {
+      vehicle_type: tripStartEvent.vehicle_type,
+      trip_id,
+      device_id,
+      provider_id,
+      start_time: tripStartEvent.timestamp,
+      end_time: tripEndEvent.timestamp,
+      start_service_area_id: tripStartEvent.service_area_id,
+      end_service_area_id: tripEndEvent.service_area_id
+    }
+    const tripTelemetry = tripMap[trip_id]
+    const telemetry: TripTelemetry[][] = []
+    if (tripTelemetry && tripTelemetry.length > 0) {
+      for (let i = 0; i < events.length - 2; i++) {
+        const start_time = events[i].timestamp
+        const end_time = events[i + 1].timestamp
+        // Exclude event telemetry points
+        const tripSegment = tripTelemetry.filter(
+          telemetry_point => telemetry_point.timestamp > start_time && telemetry_point.timestamp < end_time
+        )
+        tripSegment.sort((a, b) => (a.timestamp > b.timestamp ? 1 : -1))
+        telemetry.push(tripSegment)
+      }
+    } else {
+      await log.error('TRIP TELEMETRY NOT FOUND')
+      throw new Error('TRIP TELEMETRY NOT FOUND')
+    }
+
+    // Calculate trip metrics
+    const duration = tripEndEvent.timestamp - tripStartEvent.timestamp
+    const distMeasure = tripStartEvent.gps ? calcDistance(telemetry, tripStartEvent.gps) : null
+    const distance = distMeasure ? distMeasure.distance : null
+    const points = distMeasure ? distMeasure.points : []
+    const violationArray = points.filter(dist => {
+      return dist > config.compliance_sla.max_telemetry_distance
+    })
+    const violation_count = violationArray.length
+    const max_violation_dist = violation_count ? Math.min(...violationArray) : null
+    const min_violation_dist = violation_count ? Math.max(...violationArray) : null
+    const avg_violation_dist = violation_count ? violationArray.reduce((a, b) => a + b) / violationArray.length : null
+
+    const tripData: TripEntry = {
+      ...baseTripData,
+      duration,
+      distance,
+      violation_count,
+      max_violation_dist,
+      min_violation_dist,
+      avg_violation_dist,
+      events,
+      telemetry
+    }
+
+    await db.insertTrips(tripData)
+    // Delete all processed telemetry data and update cache
+    delete tripMap[trip_id]
+    await cache.writeTripsTelemetry(`${provider_id}:${device_id}`, tripMap)
+    return trip_id
+  }
+  await log.error('TELEMETRY NOT FOUND')
+  throw new Error('TELEMETRY NOT FOUND')
+}
+
+export async function tripAggregator() {
+  const curTime = new Date().getTime()
+  const tripsMap = await cache.readAllTripsEvents()
+  if (!tripsMap) {
+    log.info('NO TRIP EVENTS FOUND')
+    return
+  }
+  await Promise.all(

```

Comment:
Open request to anyone who thinks they can quickly clean up these nested promises to chime in... I poke around it for about an hour and it proved pretty challenging. Because this is a CRON triggered function, I'm not too worried about performance, but ¯\\_(ツ)_/¯